src/main/java/io/cdap/plugin/gcp/bigquery/relational/BigQueryRelation.java

```diff
@@ -1,117 +1,351 @@
 package io.cdap.plugin.gcp.bigquery.relational;
 
+import com.google.common.annotations.VisibleForTesting;
 import io.cdap.cdap.etl.api.aggregation.DeduplicateAggregationDefinition;
 import io.cdap.cdap.etl.api.aggregation.GroupByAggregationDefinition;
 import io.cdap.cdap.etl.api.relational.Expression;
+import io.cdap.cdap.etl.api.relational.InvalidRelation;
 import io.cdap.cdap.etl.api.relational.Relation;
+import io.cdap.plugin.gcp.bigquery.sqlengine.BigQuerySQLDataset;
+import io.cdap.plugin.gcp.bigquery.sqlengine.builder.BigQueryDeduplicateSQLBuilder;
+import io.cdap.plugin.gcp.bigquery.sqlengine.builder.BigQueryGroupBySQLBuilder;
+import io.cdap.plugin.gcp.bigquery.sqlengine.builder.BigQueryNestedSelectSQLBuilder;
+import io.cdap.plugin.gcp.bigquery.sqlengine.builder.BigQuerySelectSQLBuilder;
 
-import java.util.HashMap;
+import java.util.Collection;
+import java.util.LinkedHashMap;
 import java.util.Map;
 import java.util.Set;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
 
 /**
  * An implementation of {@link Relation} designed to operate on BigQuery.
  */
 public class BigQueryRelation implements Relation {
-  private final boolean isValid;
-  private final String validationError;
+  private static final SQLExpressionFactory factory = new SQLExpressionFactory();
 
-  private final Map<String, Expression> columns;
-  private final Expression filter;
-  private final Relation baseRelation;
   private final String datasetName;
+  private final Set<String> columns;
+  private final BigQuerySQLDataset sourceDataset;
+  private final BigQueryRelation parent;
+  private final String transformExpression;
 
   /**
-   * Initializes a new {@link BigQueryRelation} using the set of columns specified.
-   * @param columnSet set of strings indicating column names in the relation.
+   * Gets a new BigQueryRelation instance
+   *
+   * @param bqProject     Project where the source SQL table is stored
```

Comment:
Won't sourceDataset has it? Another question is that, are you sure that sourceDataset would be available when you make the relation? It's pretty early in the game, there is no job yet. At this point we did not even decide if we will do it in BQ

Reply:
I've refactored the code to supply the source datasets once the engine has all of them available to execute.